### ISSUE TITLE
Pros/cons of Elmish vs plain React components (via Fable.React)

### ISSUE THREAD

**Issue by SirUppyPancakes** [elmish/elmish#154]
I apologize if this is not the right place to ask this sort of thing (I would have asked it on the Gitter channel, but decided that it might be better if it had more visibility).

After actually starting to lay out the basis for an application (having sub-pages, tabs, and lots of nested modular views), I am starting to run up against a common problem introduced by Elm/Elmish related to the Parent-Child pattern (https://elmish.github.io/elmish/parent-child.html). Here is an article detailing the issue and some various approaches to getting around it: https://medium.com/@_rchaves_/child-parent-communication-in-elm-outmsg-vs-translator-vs-nomap-patterns-f51b2a25ecb1

Ultimately, it comes down to state. Every parent view needs to know about its child views' internal states and messages, even when it does not need to manipulate them in any way. This becomes particularly painful when you consider "custom controls". I am currently using components from [Material UI](https://material-ui.com/) which is great and has many components ready to use right at the start. However, at some point, I need to create custom controls or groups of controls that are easy re-use.

My solution to this was, upon first encountering it, was to change the particular component I was making that needed to maintain and manage its own state (without getting its parent involved) a plain React component (via Fable.React). However, it wasn't long until I ran into the same thing again and again.

At this point, I have stepped back and am wondering what I am gaining by using Elmish vs plain React components (where I can control, on a per-component basis, statefulness). I understand that the benefits of using Elm/Elmish are similar to using something like Redux to manage all state, which has cool things like a [time travelling debugger](https://medium.com/the-web-tub/time-travel-in-react-redux-apps-using-the-redux-devtools-5e94eba5e7c0), but past that I am not sure what else I might gain/lose by switching to plain React components.

TL:DR: Considering Elm vs React is usually met with "Elm provides safety at compile time, and React is more flexible", but in the context of Fable, both Elmish and Fable.React are type-safe because of F#, so what would be the additional pros/cons of using one over the other?

**Comment by MangelMaxime** [comment]
:warning: :warning: :warning: :warning:  First things first, I am probably a lot biased because I never liked pure React, Angular, Template based UI even when I was working fully in JavaScript. So the next part of my message are purely my opinion on the subject and reflect my way of using/thinking with Elmish. :warning: :warning: :warning: 

- It's easy to follow the flow of your application, you can see it directly in your model domain.
- All your application use the same structure where in React you need to consider a lot of different things. Make sure to correctly bind your context.
- Everything in Elmish is using the same primitive blocks => Everything is a function
- When using Elmish maintaining a unique state of your app is easier, when trying React in the past I never got pass the minimal application because of the store management which was (perhaps is still) a mess
- Elmish force you to implement clear communication channels between the components. Parents communicates with the child using the child function arguments or if really necessary can generate a `Msg`. Child can communicate with the parent using `ExternalMsg` and all the time you add a new message the compiler will have your back.

Another point to note, is with Elmish you can include a stateful components in your application when it make sense. For example, in Fable REPL, the editors are stateful components. In Fulma documentation, all the showcases are using stateful components as a container.

If I have to be honest in whatever application you are the time travel features is almost all the time a toy. I mean yes it make people eyes sparkles but I never use it when developing. 

However, being able to see that one cmd generates all this components in return make more sense because you can easily see if you are doing something wrong.

Yes I don't have much in favor point for a pure React approach. I really like having an Elmish based application and just be able to include React components when I need into the application. Most of the is for wrapping a complex JavaScript feature or library:
- Code editor
- Clipboard management

Another benefit I find to Elmish is it prevents you from doing crazy access to resources. For example, in recharts (charts library) every is implicit because the children of a component can access it's parents context or vice-versa to do some calculation on certain circumstance. People do that because JavaScript allow you to do it. When if your model is well designed you can do anything you want in a clear way.

**Comment by et1975** [comment]
There's two ways I'd approach this:

* Do you have view with some logic, or do you have logic with some views?
* Compile-time safety  - does it extend to runtime interactions?

Elmish lets you focus on logic first, and views second. It's super-simple to reason about what it actually *does*, as opposed to what it *looks like*.
If you are the only one working on an app, can keep everything in your head and don't have to worry that the guy inheriting this maybe a psychopath who knows where you live, then sure - go with what's comfortable ;)

Compile-time safety is often misunderstood, especially if coming from languages that don't have any. Elm and F# go beyond "int vs string" stuff into the territory of "make invalid states unrepresentable".
In other words, it's not only about static structure (does it have the shape I expect?), but also about behavior (can I access it this way at this particular moment?). The state of a child is not really "internal". The distinction I claim - is it "observable", i.e. can I reconstruct the look and feel of you component if I give it this state as an input? This is what makes the UI "unidirectional" and easy to reason about. 

All the verbosity is there to ensure the *behavior* (ie, the nature of interaction between your components) is correct by the time your program compiles! React+F# guarantees in this respect are somewhat weaker than Elm's, but still pretty amazing. When you read about experiences like "We went into production with 10K lines of code and found ZERO bugs since" - it's not an exaggeration, it's the compiler, forcing you to think about every eventuality.

We've built some fairly sophisticated UIs with [react-toolbox](http://react-toolbox.io/#/) ([bindings](https://github.com/Prolucid/fable-react-toolbox)) at prolucid, but only once have we needed to go with a stateful component - when using D3, and only because it wasn't built for React or to be stateless.

**Comment by SirUppyPancakes** [comment]
There's definitely some interesting points about the verbosity making more guarantees about correctness of intentions. I think I will try to stick it out a bit with Elmish before I toss it out completely. Being the only developer on what will turn into quite a large project, I think the biggest deciding factor will end up being how much time/energy I can spend to get that safety before my timelines start to suffer. Thanks for the input guys 😃 
Feel free to close this whenever.

**Comment by SirUppyPancakes** [comment]
Just want to add some notes here for anyone else who stumbles across this post with the same questions. After modeling things out with Elmish and then pure React, it's obvious that both serve important purposes when it comes to a complicated application with lots of state. At this point, I've settled on a hybrid approach under the following design guidelines:

1) Separate the `Model` and `Msg` types into `AppModel` + `ViewModel` and `AppMsg` + `ViewMsg`. This allows you to keep all of the view-specific state and messages separate which avoids a lot of headaches about which state matters to what:
```
type Model = {
    appModel: AppModel
    viewModel: ViewModel
}

type Msg =
    | AppMsg of AppMsg
    | ViewMsg of ViewMsg
```
Then, when deciding on whether to add state to `AppModel` or `ViewModel`, you can simply ask whether or not the state exclusively serves a specific UI concept, or if it could be expressed as any UI element or none at all. For example, whether or not a `Drawer` is open would be `ViewModel` state, since it only makes sense in the context of the `Drawer`, but an `Account` record is vital to the application logic, and could be displayed in any UI or none at all, so it belongs to `AppModel`.

2) Whenever you have a reusable component that doesn't need to expose its state or messages (they can sometimes get complicated and just pollute the `ViewModel`), use a stateful React component instead. You can still control any important state through `Props` like normal and pass the `dispatch` function if needed as a `Prop` to do messaging.

3) Have all `view` functions take a top-level `Model` and `Msg`, even if it doesn't look like that component will need data or the ability to dispatch higher up the tree. Then, you can simply create local variables for dispatching or accessing common model/msg sub-trees:
```
let view model dispatch =
    let appleMsg = dispatch << AppMsg << AppleMsg
    let addApple = appleMsg  << Add
    let editApple = appleMsg  << Edit
    let allSeeds = model.appModel.apples |> Seq.collect (fun apple -> apple.seeds)
// view code...
```
I find that if I instead try to pass only the sub-tree of `Model` or `Msg` that I need, I forget that I actually need some stuff higher up later. This leads to refactoring that involves then passing the larger sub-tree and then creating these local shortcuts inside the function anyway. This started happening multiple times and got really frustrating fast. If you always pass `Model` and `Msg`, you'll only ever have to add to your already-existing local shortcut variables which is much faster.

I'm sure there is a lot more you can do to keep things clean and tidy with Elmish, but these specific things have definitely made my situation a lot less painful and quite pleasant. Hope it helps others who come across the same issues! 😁

**Comment by et1975** [comment]
I can't agree with your recommendations, but thanks for sharing!

**Comment by tenigma** [comment]
I am not programming front-end but I have read/watched a lot about elm and I would recommend watching this [video from Richard Feldman at the Elm Europe conference](https://www.youtube.com/watch?v=DoA4Txr4GUs)
